controller/controller_server/utils.py

```diff
@@ -14,36 +16,140 @@
 logger = get_stdout_logger()
 
 
-def get_array_connection_info_from_secret(secrets):
-    user = secrets[config.SECRET_USERNAME_PARAMETER]
-    password = secrets[config.SECRET_PASSWORD_PARAMETER]
-    array_addresses = secrets[config.SECRET_ARRAY_PARAMETER].split(",")
+def _is_topology_match(secret_topologies, node_topologies):
+    is_match = False
+    for topologies in secret_topologies:
+        logger.debug(
+            "Comparing topologies: object topologies: {},"
+            " node topologies: {}".format(topologies, node_topologies))
+        if all(item in node_topologies.items() for item in topologies.items()):
+            is_match = True
+    return is_match
+
+
+def get_volume_topologies(request):
+    if request.HasField(config.REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD):
+        accessibility_requirements = request.accessibility_requirements
+        if accessibility_requirements.preferred:
+            topologies = accessibility_requirements.preferred[0].segments
+            logger.info("Chosen volume topologies: {}".format(topologies))
+            return topologies
+
+
+def get_secret_by_topologies(secret_config_string, node_topologies):
+    secret_config = ast.literal_eval(secret_config_string)
+    for secret_uid, secret_info in secret_config.items():
+        secret_topologies = secret_info.get(config.SECRET_SUPPORTED_TOPOLOGIES_PARAMETER)
+        if _is_topology_match(secret_topologies, node_topologies):
+            return secret_info, secret_uid
+
+
+def get_secret_by_uid(secret_config_string, secret_uid):
+    secret_config = ast.literal_eval(secret_config_string)
+    return secret_config.get(secret_uid)
+
+
+def get_pool_by_uid(pools, uid):
+    pools = ast.literal_eval(pools)
+    return pools.get(uid)
+
+
+def _get_secret_from_secrets(secrets, topologies=None, secret_uid=None):
+    secret_config = secrets.get(config.SECRET_CONFIG_PARAMETER)
+    secret = secrets
+    if secret_config:
+        if secret_uid:
+            secret = get_secret_by_uid(secret_config_string=secret_config, secret_uid=secret_uid)
+        elif topologies:
+            secret, secret_uid = get_secret_by_topologies(secret_config_string=secret_config,
+                                                          node_topologies=topologies)
+        else:
+            raise ValidationException(messages.invalid_secret_message)
+    return secret, secret_uid
+
+
+def _get_array_connection_info_from_secret(secret):
+    user = secret[config.SECRET_USERNAME_PARAMETER]
+    password = secret[config.SECRET_PASSWORD_PARAMETER]
+    array_addresses = secret[config.SECRET_ARRAY_PARAMETER].split(",")
     return user, password, array_addresses
 
 
-def get_volume_id(new_volume):
-    return _get_object_id(new_volume)
+def get_array_connection_info_from_secrets(secrets, topologies=None, secret_uid=None):
```

Comment:
I'm not a big fan of functions that return tuples in Python because whenever we want to change the tuple, we need to change lots of places and QA should test it.
I suggest returning a dictionary and save the keys in "constant" variables (CAP_LETTER_VARIABLE) or creating a data class for that.

Reply:
done